src/lib/ReactCore_Internal.re

```diff
@@ -337,32 +378,44 @@ module Make = (Implementation: HostImplementation) => {
       | Some(opaqueInstance) =>
         /* Throwaway update log: this is a render so no need to keep an update log. */
         let updateLog = UpdateLog.create();
-        update(~updateLog, opaqueInstance, element)
+        update(~updateLog, opaqueInstance, element);
       | None =>
-        let instance =
-          createInstance(
-            ~component,
-            ~element,
-            ~instanceSubTree=IFlat([]),
-            ~subElements=Flat([])
-          );
-        let self = createSelf(~instance);
+        let id = GlobalState.instanceIdCounter^;
+        incr(GlobalState.instanceIdCounter);
+        let (self, pendingStateUpdates) = createInitialSelf(~component);
         if (component.didMount !== defaultDidMount) {
-          component.didMount(self)
+          component.didMount(self);
         };
         switch component.elementType {
```

Comment:
We need to extract this branching somewhere. It's completely redundant and there's only one difference.